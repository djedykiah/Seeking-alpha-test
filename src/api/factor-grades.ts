import { useQuery } from '@tanstack/react-query';
import { http } from '../lib';

type FactorGradesResponse = {
  [key: string]: {
    current: string
  }
}

export const useFactorGradesQuery = (interval: 'now' | '3m' | '6m') => useQuery<FactorGradesResponse>({
  queryKey: ['factor-grades', interval],
  queryFn: async () => {
    const { data } = await http.get<FactorGradesResponse>(`/factor-grades/${interval}`);

    return data;
  },
});
