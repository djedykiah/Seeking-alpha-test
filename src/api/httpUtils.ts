const BASE_URL = 'http://seekingalpha.free.beeceptor.com';

export const request = async (url: string) => {
  const response = await fetch(`${BASE_URL}${url}`);

  if (!response.ok) {
    throw new Error(
      `Network response was not ok - status  : ${response.status}`,
    );
  }

  return response.json();
};
