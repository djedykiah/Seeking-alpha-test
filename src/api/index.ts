export * from './factorGrades';
export * from './quantRanking';
export * from './ratingsSummary';
export * from './user';
