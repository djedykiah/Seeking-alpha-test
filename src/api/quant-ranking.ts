import { useQuery } from '@tanstack/react-query';

import { http } from '../lib';

type Ranking = {
  rank: number;
  total: number;
};

export type QuantRankingResponse = {
  sector: string;
  industry: string;
  rankings: {
    industry_specific: Ranking;
    overall: Ranking;
    sector: Ranking;
  };
};

export type RankingType = keyof QuantRankingResponse['rankings'];

export const useQuantRankingQuery = () =>
  useQuery<QuantRankingResponse>({
    queryKey: ['quant-ranking'],
    queryFn: async () => {
      const { data } = await http.get<QuantRankingResponse>('/quant-ranking');

      return data;
    },
  });
