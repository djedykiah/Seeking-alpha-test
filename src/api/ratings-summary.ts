import { useQuery } from '@tanstack/react-query';
import { http } from '../lib';

type RatingsSummaryResponse = {
  [key: string]: {
    rating: 'HOLD' | 'BUY',
    score: number
  }
}

export const useRatingsSummaryQuery = () => useQuery<RatingsSummaryResponse>({
  queryKey: ['ratings-summary'],
  queryFn: async () => {
    const { data } = await http.get<RatingsSummaryResponse>('/ratings-summary');

    return data;
  },
});
