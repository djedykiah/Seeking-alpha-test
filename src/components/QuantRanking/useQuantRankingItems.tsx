import { ReactNode, useMemo } from 'react';

import {
  QuantRankingResponse,
  RankingType,
  useQuantRankingQuery,
} from '../../api';
import { Link } from '../Link';

import { getRankingTitle } from './getRankingTitle';

type QuantRankingItem = {
  title: string;
  value: ReactNode;
};

type Api = {
  items: QuantRankingItem[];
  error: Error | null;
  isLoading: boolean;
};

const getRankingsTypes = (
  rankings: QuantRankingResponse['rankings'],
): RankingType[] => Object.keys(rankings) as RankingType[];

export const useQuantRankingItems = (): Api => {
  const { data, isLoading, error } = useQuantRankingQuery();

  const items: QuantRankingItem[] = useMemo(() => {
    if (!data) {
      return [];
    }

    return [
      {
        title: 'Sector',
        value: <Link to={`/sector/${data.sector}`}>{data.sector}</Link>,
      },
      {
        title: 'Industry',
        value: <Link to={`/industry/${data.industry}`}>{data.industry}</Link>,
      },
      ...getRankingsTypes(data.rankings).map((ranking) => ({
        title: getRankingTitle(ranking),
        value: (
          <Link to={`/ranking/${ranking}`}>
            <strong>{data.rankings[ranking].rank}</strong> out of{' '}
            <strong>{data.rankings[ranking].total}</strong>
          </Link>
        ),
      })),
    ];
  }, [data]);

  return {
    items,
    isLoading,
    error,
  };
};
