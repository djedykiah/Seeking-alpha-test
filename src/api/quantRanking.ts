import { useQuery } from '@tanstack/react-query';

import { request } from './httpUtils';

type Ranking = {
  rank: number;
  total: number;
};

export type QuantRankingResponse = {
  sector: string;
  industry: string;
  rankings: {
    industry_specific: Ranking;
    overall: Ranking;
    sector: Ranking;
  };
};

export type RankingType = keyof QuantRankingResponse['rankings'];

export const useQuantRankingQuery = () =>
  useQuery<QuantRankingResponse>({
    queryKey: ['quant-ranking'],
    queryFn: () => request('/quant-ranking'),
  });
