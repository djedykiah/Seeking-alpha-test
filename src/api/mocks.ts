import { FactorGradesResponse } from './factor-grades';
import { QuantRankingResponse } from './quant-ranking';
import { RatingsSummaryResponse } from './ratings-summary';
import { UserResponse } from './user';

export const fg0: FactorGradesResponse = {
  Key1: {
    current: 'F',
  },
  Key2: {
    current: 'F',
  },
  Key3: {
    current: 'F',
  },
  Key4: {
    current: 'F',
  },
  Key5: {
    current: 'F',
  },
};

export const fg3: FactorGradesResponse = {
  Key1: 'B',
  Key2: 'B',
  Key3: 'B',
  Key4: 'B',
  Key5: 'B',
};

export const fg6: FactorGradesResponse = {
  // @ts-ignore
  data: [
    ['Key1', 'С'],
    ['Key2', 'С'],
    ['Key3', 'С'],
    ['Key4', 'С'],
    ['Key5', 'С'],
  ],
};

export const fgs: any = {
  now: fg0,
  '3m': fg3,
  '6m': fg6,
};

export const rs: RatingsSummaryResponse = {
  Key1: {
    rating: 'HOLD',
    score: 20,
  },
  Key2: {
    rating: 'BUY',
    score: 12,
  },
  Key3: {
    rating: 'BUY',
    score: 12,
  },
};

export const qr: QuantRankingResponse = {
  industry: 'Industry',
  sector: 'Sector',
  rankings: {
    overall: {
      rank: 825,
      total: 44432,
    },
    sector: {
      rank: 105,
      total: 522,
    },
    industry_specific: {
      rank: 8,
      total: 28,
    },
  },
};

export const usr: UserResponse = {
  premium: true,
};
