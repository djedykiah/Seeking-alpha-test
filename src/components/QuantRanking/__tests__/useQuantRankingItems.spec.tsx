import { renderHook } from '@testing-library/react';
import { Mock } from 'vitest';

import { useQuantRankingQuery } from '../../../api';
import { useQuantRankingItems } from '../useQuantRankingItems';

vi.mock('../../../api', () => ({
  useQuantRankingQuery: vi.fn(),
}));

vi.mock('../getRankingTitle', () => ({
  getRankingTitle: vi
    .fn()
    .mockImplementation((ranking: string) => `Title for ${ranking}`),
}));

describe('useQuantRankingItems', () => {
  it('should return loading state', () => {
    (useQuantRankingQuery as Mock).mockReturnValue({
      data: null,
      isLoading: true,
      error: null,
    });

    const { result } = renderHook(() => useQuantRankingItems());

    expect(result.current.isLoading).toBe(true);
    expect(result.current.error).toBeNull();
    expect(result.current.items).toEqual([]);
  });

  it('should return error state', () => {
    const error = new Error('Failed to fetch');
    (useQuantRankingQuery as Mock).mockReturnValue({
      data: null,
      isLoading: false,
      error,
    });

    const { result } = renderHook(() => useQuantRankingItems());

    expect(result.current.isLoading).toBe(false);
    expect(result.current.error).toBe(error);
    expect(result.current.items).toEqual([]);
  });

  it('should return items when data is available', () => {
    const mockData = {
      sector: 'Tech',
      industry: 'Software',
      rankings: {
        type1: { rank: 1, total: 10 },
        type2: { rank: 5, total: 20 },
      },
    };
    (useQuantRankingQuery as Mock).mockReturnValue({
      data: mockData,
      isLoading: false,
      error: null,
    });

    const { result } = renderHook(() => useQuantRankingItems());

    expect(result.current.isLoading).toBe(false);
    expect(result.current.error).toBeNull();

    expect(result.current.items).toMatchSnapshot();
  });

  it('should handle case when data is undefined', () => {
    (useQuantRankingQuery as Mock).mockReturnValue({
      data: undefined,
      isLoading: false,
      error: null,
    });

    const { result } = renderHook(() => useQuantRankingItems());

    expect(result.current.isLoading).toBe(false);
    expect(result.current.error).toBeNull();
    expect(result.current.items).toEqual([]);
  });
});
