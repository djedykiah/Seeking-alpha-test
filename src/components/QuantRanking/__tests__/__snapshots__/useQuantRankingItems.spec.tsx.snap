// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useQuantRankingItems > should return items when data is available 1`] = `
[
  {
    "title": "Sector",
    "value": <Link
      to="/sector/Tech"
    >
      Tech
    </Link>,
  },
  {
    "title": "Industry",
    "value": <Link
      to="/industry/Software"
    >
      Software
    </Link>,
  },
  {
    "title": "Title for type1",
    "value": <Link
      to="/ranking/type1"
    >
      <strong>
        1
      </strong>
       out of
       
      <strong>
        10
      </strong>
    </Link>,
  },
  {
    "title": "Title for type2",
    "value": <Link
      to="/ranking/type2"
    >
      <strong>
        5
      </strong>
       out of
       
      <strong>
        20
      </strong>
    </Link>,
  },
]
`;
